pallets/domains/src/lib.rs

```diff
@@ -0,0 +1,517 @@
+        /// Sets the domain content to be an outside link.
+        #[pallet::weight(<T as Config>::WeightInfo::set_domain_content())]
+        pub fn set_domain_content(
+            origin: OriginFor<T>,
+            domain: DomainName<T>,
+            new_content: Content,
+        ) -> DispatchResult {
+            let sender = ensure_signed(origin)?;
+
+            let domain_lc = Self::lower_domain_then_bound(domain.clone());
+            let mut meta = Self::require_domain(domain_lc.clone())?;
+
+            Self::ensure_allowed_to_update_domain(&meta, &sender)?;
+
+            ensure!(meta.content != new_content, Error::<T>::DomainContentNotChanged);
+            ensure_content_is_valid(new_content.clone())?;
+
+            meta.content = new_content;
+            RegisteredDomains::<T>::insert(&domain_lc, meta);
+
+            Self::deposit_event(Event::DomainUpdated(sender, domain));
+            Ok(())
+        }
+
+        /// Mark set of domains as not reservable by users.
+        #[pallet::weight(<T as Config>::WeightInfo::reserve_domains(T::DomainsInsertLimit::get()))]
+        pub fn reserve_domains(
+            origin: OriginFor<T>,
+            domains: Vec<DomainName<T>>,
+        ) -> DispatchResultWithPostInfo {
+            ensure_root(origin)?;
+
+            let domains_len = domains.len();
+            Self::ensure_domains_insert_limit_not_reached(domains_len)?;
+
+            let inserted_domains_count = Self::insert_domains(
+                domains,
+                Self::ensure_valid_domain,
+                |domain| ReservedDomains::<T>::insert(domain, true),
+            )?;
+
+            Self::deposit_event(Event::DomainsReserved(domains_len as u16));
+            Ok((
+                Some(<T as Config>::WeightInfo::reserve_domains(inserted_domains_count)),
+                Pays::No,
+            ).into())
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        fn do_register_domain(
+            owner: T::AccountId,
+            full_domain: DomainName<T>,
+            content: Content,
+            expires_in: <T as frame_system::pallet::Config>::BlockNumber,
+            reserve_deposit: ReserveDeposit,
+        ) -> DispatchResult {
+            ensure!(!expires_in.is_zero(), Error::<T>::ZeroReservationPeriod);
+            ensure!(
+                expires_in <= T::RegistrationPeriodLimit::get(),
+                Error::<T>::TooBigRegistrationPeriod,
+            );
+
+            // Note that while upper and lower case letters are allowed in domain
+            // names, domain names are not case-sensitive. That is, two names with
+            // the same spelling but different case are to be treated as if identical.
+            let domain_lc = Self::lower_domain_then_bound(full_domain.clone());
+
+            ensure!(!Self::is_domain_reserved(&domain_lc), Error::<T>::DomainIsReserved);
+
+            ensure_content_is_valid(content.clone())?;
+
+            Self::ensure_valid_domain(&domain_lc)?;
```

Comment:
Better call this `ensure_valid_domain()` before we accessing storage in `is_domain_reserved()`. It can save some performance for us, because accessing a blockchain storage is an expensive operation.

Reply:
As you wish. But FYI:

- 1 storage read is of `25_000_000` weight constantly.
- lowering the smallest domain is of `10_000_000` weight on the 16 core CPU and 32 GB RAM.